File: examples/go/clean-api-go/domain/repositories/sale_repository.go
package repositories

import (
	"context"

	"github.com/freakjazz/clean-api-go/domain/entities"
)

type ISaleRepository interface {
	Save(ctx context.Context, sale *entities.Sale) (*entities.Sale, error)
	FindByProductID(ctx context.Context, productID string, page, pageSize int) ([]*entities.Sale, int, error)
}

File: examples/python/clean_api_python/domain/repositories/sale_repository.py
from abc import ABC, abstractmethod
from typing import List, Optional, Tuple

from ..entities.sale import Sale


class ISaleRepository(ABC):
    @abstractmethod
    def save(self, sale: Sale) -> Sale: ...

    @abstractmethod
    def find_by_product_id(
        self, product_id: Optional[str], page: int = 1, page_size: int = 10
    ) -> Tuple[List[Sale], int]: ...
